code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/CarouselTemplateUtil.kt

```diff
@@ -0,0 +1,115 @@
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.services.ui.notification
+
+import android.content.Intent
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.services.ui.notification.models.CarouselPushTemplate
+import org.json.JSONArray
+import org.json.JSONException
+
+internal object CarouselTemplateUtil {
+    private const val SELF_TAG = "CarouselTemplateHelpers"
+
+    /**
+     * Calculates a new left, center, and right index given the current center index, total number
+     * of images, and the intent action.
+     *
+     * @param centerIndex [Int] containing the current center image index
+     * @param listSize `Int` containing the total number of images
+     * @param action [String] containing the action found in the broadcast [Intent]
+     * @return [List] containing the new calculated left, center, and right indices
+     */
+    internal fun calculateNewIndices(
+        centerIndex: Int,
+        listSize: Int?,
+        action: String?
+    ): List<Int> {
+        if (listSize == null || listSize < CarouselPushTemplate.MINIMUM_FILMSTRIP_SIZE) return emptyList()
+        val newIndices = mutableListOf<Int>()
+        var newCenterIndex = 0
+        var newLeftIndex = 0
+        var newRightIndex = 0
+        Log.trace(
+            PushTemplateConstants.LOG_TAG,
+            SELF_TAG,
+            "Current center index is $centerIndex and list size is $listSize."
+        )
+        if ((action == PushTemplateConstants.IntentActions.FILMSTRIP_LEFT_CLICKED) || (action == PushTemplateConstants.IntentActions.MANUAL_CAROUSEL_LEFT_CLICKED)) {
+            newCenterIndex = (centerIndex - 1 + listSize) % listSize
+            newLeftIndex = (newCenterIndex - 1 + listSize) % listSize
+            newRightIndex = centerIndex
+        } else if ((
+            (action == PushTemplateConstants.IntentActions.FILMSTRIP_RIGHT_CLICKED) || (
+                action ==
+                    PushTemplateConstants.IntentActions.MANUAL_CAROUSEL_RIGHT_CLICKED
+                )
+            )
+        ) {
+            newCenterIndex = (centerIndex + 1) % listSize
+            newLeftIndex = centerIndex
+            newRightIndex = (newCenterIndex + 1) % listSize
+        }
+        newIndices.add(newLeftIndex)
+        newIndices.add(newCenterIndex)
+        newIndices.add(newRightIndex)
+        Log.trace(
+            PushTemplateConstants.LOG_TAG,
+            SELF_TAG,
+            "Calculated new indices. New center index is $newCenterIndex, new left index is $newLeftIndex, and new right index is $newRightIndex."
+        )
+        return newIndices
+    }
+
+    internal fun parseCarouselItems(carouselItemsString: String?): MutableList<CarouselPushTemplate.CarouselItem> {
```

Comment:
Is this being used anywhere outside of `CarouselPushTemplate`. Can we make this a private companion method for that class if not.